### ISSUE TITLE
Refresh Token error - SecretHash does not match for the client

### ISSUE THREAD

**Issue by w5922xd** [aws/aws-sdk-net-extensions-cognito#122]
### Describe the bug

When calling `StartWithRefreshTokenAuthAsync`, I am receiving an error `Amazon.CognitoIdentityProvider.Model.NotAuthorizedException: 'SecretHash does not match for the client: xxxxxxxxxxxx`. 

I have double-checked the calculation of the SecretHash and it looks correct for me. When I use the SRP AuthFlow it works just fine using the exact same SecretHash value. I captured the web traffic for the `POST` request sent to the api and the request syntax matches the `InitiateAuth` [documentation](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_InitiateAuth.html). 

I can see that others have reported this problem in this repo [here](https://github.com/aws/aws-sdk-net-extensions-cognito/issues/94) and [here](https://github.com/aws/aws-sdk-net-extensions-cognito/issues/35). I looked through it and used [this code](https://github.com/aws/aws-sdk-net-extensions-cognito/issues/35#issuecomment-712406376) to try and reproduce the bug. I get the same error regarding the SecretHash.

### Expected Behavior

I am expecting a response from the API. 

### Current Behavior

I am receiving an error `Amazon.CognitoIdentityProvider.Model.NotAuthorizedException: 'SecretHash does not match for the client: xxxxxxxxxxxx`. 

### Reproduction Steps

I looked through it and used [this code](https://github.com/aws/aws-sdk-net-extensions-cognito/issues/35#issuecomment-712406376) to try and reproduce the bug. I get the same error regarding the SecretHash.

### Additional Information/Context

I also have seen a [comment on stack overflow](https://stackoverflow.com/a/72100400) that instead of user email being used in the secret hash you actually need to use the sub value. In this case the SecretHash value differs from the SRP auth flow but the result is the same: 'SecretHash does not match for the client:`. 

### AWS .NET SDK and/or Package version used

Amazon.Extensions.CognitoAuthentication v2.4.2

### Targeted .NET Platform

.net6

### Operating System and version

Windows 10

**Comment by w5922xd** [comment]
Hello @ashishdhingra. Apologies for the delayed response. 

* Sample code to reproduce the issue.
I have reproduced this problem with the code shared in https://github.com/aws/aws-sdk-net-extensions-cognito/issues/94#issuecomment-1267336763

Here is my output: 
```
Enter User Pool ID: us-west-2_<<removed>>
Enter Client ID: 2nm<<removed>>
Enter Client Secret: c7iq<<removed>>
User Name: <<removed>>
Do you have a Refresh Token (Y/N): n

Password: <<removed>>

Executing Username/Password auth flow.
ID Token: <<removed>>
Access Token <<removed>>
Refresh Token: <<removed>>

Re-execute Flow (Y/N): y
User Name: <<removed>>
Do you have a Refresh Token (Y/N): y

Existing Refresh Token: eyJjd<<removed>>

Executing Refresh Token auth flow.
```

I then error with this exception: `Amazon.CognitoIdentityProvider.Model.NotAuthorizedException: 'SecretHash does not match for the client: 2nm<<removed>>`. 

* How are AWS credentials configured?

At the moment I am just reading them in from a file and/or setting them in the code because I'm just trying to prototype a working example. I have no issues with SRP authflow and it uses the same SecretHash.

* How is App Client configured? Is it configured with client secret?
I'm not sure how to answer this exactly. It's the same setup as linked in issue 94 if that helps at all. It is configured with a client secret.

* What are the sign-in options configured for user pool?

Coginto user pool sign-in options: User name, Email
Multi-factor authentication: Optional MFA (not currently configured for my accounts)
Self-service account recovery: Enabled

**Comment by ashishdhingra** [comment]
@w5922xd The code shared in https://github.com/aws/aws-sdk-net-extensions-cognito/issues/94#issuecomment-1267336763 works fine at my end. I suspect this has to do something with your user pool and/or app client settings.  Please share the screenshots of the settings (masking the sensitive details) and the sample code solution to reproduce the issue.

Thanks,
Ashish

**Comment by w5922xd** [comment]
Hello @ashishdhingra and thanks again for looking at this problem. In the process of gathering these screenshots I believe we've uncovered our solution.

We have been logging in with our email address. This works just fine for SRP Auth flow and we get our Id, access, and refresh token. However, to redeem that refresh token, we cannot use the email address. My mistake was thinking our username and email address are the same credential since it seemed to be working fine for authentication. 

I now see this isn't true, that either email or username are acceptable for SRP auth but **NOT** for the refresh token. After making this realization  I am now able to use the refresh token and exchange it for a new set of Id, access, and refresh tokens. 

I appreciate your time spent working with me on this issue with me and apologize for any time wasted. Hopefully something within this issue will be useful to someone else. Have a great weekend.

**Comment by darcy-vitacca** [comment]
@w5922xd running into this issue, any chance you could provide code of how you are manually refreshing the token?

**Comment by w5922xd** [comment]
@darcy-vitacca I can give it a shot. I was tasked with prototyping cognito in a WPF application so this code is slapped together to show proof of concept. It might help though. 

Authenticating first with srp flow using an email address: 
```c#
public async Task<AuthFlowResponse> AuthenticateWithSrpAsync()
{            
    var userPool = new CognitoUserPool(UserPoolId, ClientId, _provider);
    var user = new CognitoUser(_username, ClientId, userPool, _provider, ClientSecret);
    
    AuthFlowResponse authResponse = await user.StartWithSrpAuthAsync(new InitiateSrpAuthRequest()
    {
        Password = _password
    }).ConfigureAwait(false);    
    return authResponse;
}
```

Then getting user attributes so I can find the actual username. This returns a tuple but the first value is the username we need to use for the refresh token.
```c#
public async Task<(string Username, List<AttributeType> Attributes)> GetUserAttributesAsync(string accessToken)
{
    var getUserRequest = new GetUserRequest()
    {
        AccessToken = accessToken
    };
    var getUserResponse = await _provider.GetUserAsync(getUserRequest);
    return (getUserResponse.Username, getUserResponse.UserAttributes);
}
```

Finally, use the refresh token. We're going to pass in the username (the value from `getUserResponse.Username`) as a param. 
```c#
public async Task<AuthFlowResponse> GetCredsFromRefreshAsync(string userId, string refreshToken)
{
    var userPool = new CognitoUserPool(UserPoolId, ClientId, _provider);
    
    var user = new CognitoUser(userId, ClientId, userPool, _provider, ClientSecret)
    {
        SessionTokens = new CognitoUserSession(null, null, refreshToken, DateTime.UtcNow, DateTime.UtcNow.AddHours(1))
    };

    var authResponse = await user.StartWithRefreshTokenAuthAsync(new InitiateRefreshTokenAuthRequest
    {
        AuthFlowType = AuthFlowType.REFRESH_TOKEN_AUTH
    });
    return authResponse;
}
```

Hope that helps.

**Comment by darcy-vitacca** [comment]
@w5922xd  Thanks for getting back to me much appreciated

**Comment by Akashjerin** [comment]
Hi,
    I am trying to auth the refreshToken. But i am getting the error "NotAuthorizedException : Secrethash does not match for the client". 
    I am getting this error for some users(7 out of 10 users in my pool i am facing the error.). And for some other users, it is working fine.
    I am using java. If my code is wrong, then how it is working for some users. 
    
    CODE :
    private static String calculateSecretHash(String clientId, String clientSecret, String userName) {
       String message = userName + clientId;
       byte[] hmac = HmacUtils.hmacSha256(clientSecret.getBytes(StandardCharsets.UTF_8), message.getBytes(StandardCharsets.UTF_8));
       return Base64.encodeBase64String(hmac);
       }
       
       
    String secretHash = calculateSecretHash(client_id, client_secret, username);
    AdminInitiateAuthRequest  authRequest = new AdminInitiateAuthRequest()
                .withClientId(client_id)
                .withAuthFlow(AuthFlowType.REFRESH_TOKEN_AUTH)
                .addAuthParametersEntry("USERNAME", username)
                .addAuthParametersEntry("REFRESH_TOKEN", refreshToken)
                .addAuthParametersEntry("SECRET_HASH", secretHash);
        AdminInitiateAuthResult authResponse;
        authResponse = cognitoClient.adminInitiateAuth(authRequest);

**Comment by w5922xd** [comment]
Hello @Akashjerin, 

I cannot help with the java SDK version unfortunately. However, I assume you found this issue by searching for the error message that the secret hash does not match. 

When I encountered that error, it was because I was using an email address instead of the username defined in the userpool. That might explain why it works for you in some instances. You **can** use the email address to sign in with certain auth flows (like SRP), but it will **not work** for getting the refresh token. You must use their username in the userpool. At least this was the case for me.

Try checking the users in your userpool. Are the ones that work making an email address their username? That was the case for me. I then realized that username and email address are separate fields. In order to get the refresh token you must supply the username in the userpool which can be parsed from the original token.

Hope that helps. Good luck!
